PC-24819: make soyjs output ES6 compatible javascript (imports/exports)

ovkulkarni opened the pull request:
- export functions that are declared (templates)
- import functions that are not already in the file and are called 
(call, print directive, func)
- switch to resolvers instead of polluting soyjs.Funcs and 
soyjs.PrintDirectives (making it easier to know when to import - defined 
by caller)
  - defaults are available as soyjs.DefaultDirectiveResolver and 
soyjs.DefaultFuncResolver

mhupman commented on soyjs/exec.go:
```diff
@@ -15,25 +15,90 @@ import (
+	outTempR, outTempW := io.Pipe()
+	var s = &state{wr: outTempW, imports: importsBuf, directiveResolver: directiveResolver, funcResolver: funcResolver, options: options}
+
+	go func() {
```
why do we need to do this in a goroutine (vs invoking directly)?

ovkulkarni replied:
the Pipe() will hang unless it is invoked in a goroutine

mhupman commented on soyjs/exec.go:
```diff
@@ -15,25 +15,90 @@ import (
-	s.walk(node)
+	importsBuf := bytes.NewBufferString("")
+	outTempR, outTempW := io.Pipe()
+	var s = &state{wr: outTempW, imports: importsBuf, directiveResolver: directiveResolver, funcResolver: funcResolver, options: options}
```
once you reach this many attributes i'd put each on a newline

mhupman commented on soyjs/exec.go:
```diff
@@ -15,25 +15,90 @@ import (
+		s.scope.push()
+		s.walk(node)
+	}()
+	fileBuf := bytes.NewBufferString("")
```
i think a `buf := &bytes.Buffer{}` works, no need for NewBufferString()

mhupman commented on .travis.yml:
```diff
@@ -9,5 +9,6 @@ install:
+  - go get github.com/robfig/soy
 script:
-  - go test github.com/robfig/soy/...
+  - go test github.com/$TRAVIS_REPO_SLUG/...
```
great find!

mhupman commented on .travis.yml:
```diff
@@ -9,5 +9,6 @@ install:
   - go get github.com/robertkrimen/otto
   - go get github.com/howeyc/fsnotify
   - go get github.com/robfig/gettext/po
+  - go get github.com/robfig/soy
```
why do we need this?

ovkulkarni replied:
the imports are in the form github.com/robfig/soy/{package}

mhupman replied:
Hmm, yeah.  Not sure Travis will work for you as most of the code uses robfig/soy/ and not ovkulkarni/soy, right?  Like you aren't testing just the code in your repo.

mhupman reviewed (commented):
I've been pondering this some more, and I can't shake the feeling that we could unify the resolving logic somehow.

```
[PREAMBLE]
[IMPORTS]
[TEMPLATES]
[EXPORTS]
```

and the function I'm thinking of now is something like

```
type CallType int

const (
  CallTypeTemplate iota
  CallTypeFunc
  CallTypeDirective
)

type ImportWriter func(w io.Writer, type CallType, name string)
```

and we'd implement something like

```
importWriter = func(w io.Writer, type CallType, name string) {
  switch type {
    case CallTypeFunc:
      w.Write([]byte("import {"+name+"}" from "js/components/Soy2JS/funcs.js"))
  }
}

soyjs.Write(..., ..., importWriter)
```

That still leaves us with the oddity of how to stitch the file together in the right order...

Lastly, we might consider adding the rewriter things to the soy2js.Options and avoid overloading the function further.  See https://github.com/ovkulkarni/soy/blob/master/soyjs/generator.go#L12

mhupman commented on soyjs/exec.go:
```diff
@@ -305,7 +381,12 @@ func (s *state) visitPrint(node *ast.PrintNode) {
 	s.js(s.bufferName, " += ")
 	for _, dir := range directives {
-		s.js(PrintDirectives[dir.Name].Name, "(")
+		d, importFile, importThis, _ := s.options.Resolver(dir.Name, CallTypeDirective)
```
dang, this turned out odd.  hmm....

mhupman reviewed (commented):
The func is kinda... funky.  I didn't think about the weird type assertion stuff you'd have to do.  We might be on the right track though, what about a tweak?  Please feel free to discuss, I'm just thinking out loud here.

```
type JSResolver interface {
  Template(name string) string
  Directive(name string) Directive, string
  Function(name string) Func, string
}

type defaultResolver struct {}

func (r defaultResolver) Template(name string) string {
  return name + ".js"
}

... and so on ...

var DefaultResolver = &defaultResolver{}
```

mhupman commented on soyjs/exec.go:
```diff
@@ -25,15 +26,116 @@ type state struct {
+}
+
+func (f ES6Formatter) Call(name string) (string, string) {
+	return importStyle(name), "import { " + importStyle(name) + " } from '" + name + ".js';\n"
```
I'd suggest not including the newline here and handling that during the join/write phase

mhupman commented on soyjs/generator.go:
```diff
@@ -8,9 +8,17 @@ import (
+type JSFormatter interface {
+	Template(name string) (string, string)
+	Call(name string) (string, string)
+	Directive(name string) (PrintDirective, string)
```
seems like the second return value is never used.. can we remove?

ovkulkarni replied:
this is used on lines 419 and 420

mhupman replied:
My point is that ES5Formatter and ES6Formatter always return empty string `""`, so it looks like it doesn't serve a purpose

mhupman replied:
e.g. importString is always ""

ovkulkarni replied:
@mhupman that is because the directives and functions declared in soyjs don't require imports. When creating our own `JSFormatter`, we want to be able to specify the import string, like `return &PrintDirective{Name: 'foo', CancelAutoescape: true}, "import { foo } from 'foo.js';"`

mhupman commented on soyjs/generator.go:
```diff
@@ -8,9 +8,17 @@ import (
+	Template(name string) (string, string)
+	Call(name string) (string, string)
+	Directive(name string) (PrintDirective, string)
+	Function(name string) (Func, string)
```
same with this param

ovkulkarni replied:
this is used on lines 443-444

mhupman commented:
Unit tests still failing?

ovkulkarni commented:
@mhupman the unit tests are failing because they are trying to test on github.com/robfig/soy, but they can't find that repository. There is a version of `.travis.yml` on the `build` branch that tests my repository, and all the tests are passing both there and locally

mhupman commented:
Great thanks re: unit tests!

mhupman commented:
Looks great!  Hopefully the tests are still passing.  I'd say were good to go to Rob's repo - can you just squash commits first so it looks clean?

Commit:
PC-24819: make soyjs output ES6 compatible javascript (imports/exports)
- export functions that are declared (templates)
- import functions that are not already in the file and are called
(call, print directive, func)
- switch to resolvers instead of polluting soyjs.Funcs and
soyjs.PrintDirectives (making it easier to know when to import - defined
by caller)
  - defaults are available as soyjs.DefaultDirectiveResolver and
soyjs.DefaultFuncResolver
